wandb/sdk/launch/sweeps/scheduler_sweep.py

```diff
@@ -27,35 +22,53 @@ class SweepScheduler(Scheduler):
     def __init__(
         self,
         *args: Any,
-        num_workers: int = 8,
-        heartbeat_queue_timeout: float = 1.0,
-        heartbeat_queue_sleep: float = 1.0,
         **kwargs: Any,
     ):
         super().__init__(*args, **kwargs)
-        self._num_workers: int = num_workers
-        # Thread will pop items off the Sweeps RunQueue using AgentHeartbeat
-        # and put them in this internal queue, which will be used to populate
-        # the Launch RunQueue
-        self._heartbeat_queue: "queue.Queue[SweepRun]" = queue.Queue()
-        self._heartbeat_queue_timeout: float = heartbeat_queue_timeout
-        self._heartbeat_queue_sleep: float = heartbeat_queue_sleep
-
-    def _start(self) -> None:
-        for worker_id in range(self._num_workers):
-            _logger.debug(f"{LOG_PREFIX}Starting AgentHeartbeat worker {worker_id}\n")
-            agent_config = self._api.register_agent(
-                f"{socket.gethostname()}-{worker_id}",  # host
-                sweep_id=self._sweep_id,
-                project_name=self._project,
-                entity=self._entity,
-            )
-            self._workers[worker_id] = _Worker(
-                agent_config=agent_config,
-                agent_id=agent_config["id"],
+
+    def _get_next_sweep_run(self, worker_id: int) -> Optional[SweepRun]:
+        """Called by the main scheduler execution loop.
+
+        Expected to return a properly formatted SweepRun if the scheduler
+        is alive, or None and set the appropriate scheduler state
+        """
+        commands: List[Dict[str, Any]] = self._get_sweep_commands(worker_id)
+        for command in commands:
+            # The command "type" can be one of "run", "resume", "stop", "exit"
+            _type = command.get("type")
+            if _type in ["exit", "stop"]:
+                run_cap = command.get("run_cap")
+                if run_cap is not None:
+                    # If Sweep hit run_cap, go into flushing state
+                    wandb.termlog(f"{LOG_PREFIX}Sweep hit run_cap: {run_cap}")
+                    self.state = SchedulerState.FLUSH_RUNS
```

Comment:
I don't love having state management occuring in `_get_next_sweep_run` directly. I feel like it makes writing this function challenging for users because they'll need to know how our state machine works

Reply:
I think we will have to have SOME state management here, for example how do we know when to stop the sweep entirely? From the main scheduler class, we will continue on forever, as long as `_get_next_sweep_run` keeps returning stuff. I moved the run_cap handling out, its not required anymore. However, users will still have to set `SchedulerState.STOPPED` to complete the sweep. I think thats not bad, perhaps not the cleanest. @KyleGoyette

Reply:
2 alternatives:
1. Add a function called "stop_sweep" that abstracts the sweep state away from users, maybe with an optional flush arg
2. If this returns None, flush the sweep and have our sweep agents check for stop in a separate function